File: skills/optimization/reference/src/bfgs.ts
/**
 * BFGS (Broyden-Fletcher-Goldfarb-Shanno) quasi-Newton optimization.
 *
 * Maintains an approximation to the inverse Hessian matrix, updated at each
 * step using gradient information. The search direction is d = -H * g, where
 * H is the inverse Hessian approximation.
 *
 * Uses Strong Wolfe line search by default (required to maintain positive-
 * definiteness of the Hessian approximation).
 *
 * @node bfgs
 * @depends-on vec-ops, result-types, line-search, finite-diff
 * @contract bfgs.test.ts
 * @hint hessian-update: The BFGS update formula is the most complex part.
 *       Translating it requires careful attention to matrix operations.
 *       In languages without matrix libraries, use explicit loops over n×n arrays.
 * @hint curvature-guard: Skip the Hessian update when yk·sk <= 0 (negative
 *       curvature). This prevents the approximation from becoming indefinite.
 * @hint line-search: BFGS requires Wolfe conditions (not just Armijo) to
 *       guarantee positive-definite updates. Using backtracking alone may
 *       cause the Hessian approximation to degenerate.
 */

import { dot, sub, scale, negate, normInf, addScaled } from "./vec-ops";
import {
  type OptimizeResult,
  type OptimizeOptions,
  defaultOptions,
  checkConvergence,
  isConverged,
  convergenceMessage,
} from "./result-types";
import { wolfeLineSearch } from "./line-search";
import { forwardDiffGradient } from "./finite-diff";

/**
 * Minimize a function using the BFGS quasi-Newton method.
 *
 * If no gradient function is provided, forward finite differences are used.
 *
 * @provenance Algorithm from Nocedal & Wright, Numerical Optimization, Chapter 6
 * @provenance Initial H0 = identity (standard choice, matches scipy and Optim.jl)
 * @provenance Curvature condition guard: skip update when yk·sk <= 0
 *            (matches Optim.jl behavior; scipy uses a damped BFGS variant instead)
 */
export function bfgs(
  f: (x: number[]) => number,
  x0: number[],
  grad?: (x: number[]) => number[],
  options?: Partial<OptimizeOptions>,
): OptimizeResult {
  const opts = defaultOptions(options);
  const n = x0.length;

  // Gradient function: analytic or finite differences
  const gradFn = grad ?? ((x: number[]) => forwardDiffGradient(f, x));

  // State
  let x = x0.slice();
  let fx = f(x);
  let gx = gradFn(x);
  let functionCalls = 1;
  let gradientCalls = 1;

  // Initialize inverse Hessian approximation as identity matrix (n×n)
  let H = identityMatrix(n);

  // Check if already at a minimum
  let gradNorm = normInf(gx);
  const initialCheck = checkConvergence(gradNorm, Infinity, Infinity, 0, opts);
  if (initialCheck && isConverged(initialCheck)) {
    return {
      x, fun: fx, gradient: gx.slice(),
      iterations: 0, functionCalls, gradientCalls,
      converged: true, message: convergenceMessage(initialCheck),
    };
  }

  for (let iteration = 1; iteration <= opts.maxIterations; iteration++) {
    // Search direction: d = -H * g
    const d = negate(matVecMul(H, gx));

    // Line search (Strong Wolfe)
    const ls = wolfeLineSearch(f, gradFn, x, d, fx, gx);
    functionCalls += ls.functionCalls;
    gradientCalls += ls.gradientCalls;

    if (!ls.success) {
      return {
        x, fun: fx, gradient: gx.slice(),
        iterations: iteration, functionCalls, gradientCalls,
        converged: false,
        message: "Stopped: line search failed to find acceptable step",
      };
    }

    // Step and gradient difference
    const xNew = addScaled(x, d, ls.alpha);
    const fNew = ls.fNew;
    const gNew = ls.gNew ?? gradFn(xNew);
    if (!ls.gNew) gradientCalls++;

    const sk = sub(xNew, x);        // step: x_{k+1} - x_k
    const yk = sub(gNew, gx);       // gradient change: g_{k+1} - g_k

    const stepNorm = normInf(sk);
    const funcChange = Math.abs(fNew - fx);
    gradNorm = normInf(gNew);

    // Update state
    x = xNew;
    fx = fNew;
    gx = gNew;

    // Check convergence
    const reason = checkConvergence(gradNorm, stepNorm, funcChange, iteration, opts);
    if (reason) {
      return {
        x: x.slice(), fun: fx, gradient: gx.slice(),
        iterations: iteration, functionCalls, gradientCalls,
        converged: isConverged(reason), message: convergenceMessage(reason),
      };
    }

    // BFGS inverse Hessian update
    const ys = dot(yk, sk);

    // Curvature condition guard: skip update if yk·sk <= 0
    if (ys <= 1e-10) {
      continue;
    }

    // H_{k+1} = (I - rho*s*y') * H_k * (I - rho*y*s') + rho*s*s'
    // where rho = 1 / (y' * s)
    const rho = 1.0 / ys;
    H = bfgsUpdate(H, sk, yk, rho);
  }

  return {
    x: x.slice(), fun: fx, gradient: gx.slice(),
    iterations: opts.maxIterations, functionCalls, gradientCalls,
    converged: false,
    message: `Stopped: reached maximum iterations (${opts.maxIterations})`,
  };
}

/** Create an n×n identity matrix as flat array of rows. */
function identityMatrix(n: number): number[][] {
  const m: number[][] = [];
  for (let i = 0; i < n; i++) {
    const row = new Array(n).fill(0);
    row[i] = 1;
    m.push(row);
  }
  return m;
}

/** Matrix-vector multiply: result = M * v. */
function matVecMul(M: number[][], v: number[]): number[] {
  const n = v.length;
  const result = new Array(n);
  for (let i = 0; i < n; i++) {
    let sum = 0;
    for (let j = 0; j < n; j++) {
      sum += M[i][j] * v[j];
    }
    result[i] = sum;
  }
  return result;
}

/**
 * BFGS inverse Hessian update.
 *
 * H_{k+1} = (I - rho*s*y') * H * (I - rho*y*s') + rho*s*s'
 *
 * @provenance Nocedal & Wright, Eq. 6.17
 */
function bfgsUpdate(
  H: number[][],
  s: number[],
  y: number[],
  rho: number,
): number[][] {
  const n = s.length;
  const Hy = matVecMul(H, y);
  const yHy = dot(y, Hy);

  const Hnew: number[][] = [];
  for (let i = 0; i < n; i++) {
    const row = new Array(n);
    for (let j = 0; j < n; j++) {
      // (I - rho*s*y') * H * (I - rho*y*s') + rho*s*s'
      // Expanded: H_ij - rho*s_i*(Hy)_j - rho*(Hy)_i*s_j + rho^2*s_i*s_j*yHy + rho*s_i*s_j
      // Simplified: H_ij - rho*(s_i*Hy_j + Hy_i*s_j) + (rho + rho^2*yHy)*s_i*s_j
      row[j] =
        H[i][j] -
        rho * (s[i] * Hy[j] + Hy[i] * s[j]) +
        rho * (1 + rho * yHy) * s[i] * s[j];
    }
    Hnew.push(row);
  }

  return Hnew;
}

File: research/experiments/sweep/runs/opt-full-python-sonnet45/src/bfgs.py
"""
Full-memory BFGS quasi-Newton optimizer.
"""

from typing import Callable, List, Optional

from .vec_ops import dot, norm_inf, sub, add_scaled, negate
from .result_types import (
    OptimizeResult, default_options, check_convergence,
    is_converged, convergence_message,
)
from .line_search import wolfe_line_search
from .finite_diff import forward_diff_gradient


def identity_matrix(n: int) -> List[List[float]]:
    return [[1.0 if i == j else 0.0 for j in range(n)] for i in range(n)]


def mat_vec_mul(M: List[List[float]], v: List[float]) -> List[float]:
    n = len(v)
    return [sum(M[i][j] * v[j] for j in range(n)) for i in range(n)]


def bfgs_update(H: List[List[float]], s: List[float], y: List[float], rho_val: float) -> List[List[float]]:
    """BFGS inverse Hessian update: H_{k+1} = (I - rho*s*y^T)*H*(I - rho*y*s^T) + rho*s*s^T"""
    n = len(s)
    # Compute H*y
    Hy = mat_vec_mul(H, y)
    # Compute y^T*H
    yTH = [sum(y[i] * H[i][j] for i in range(n)) for j in range(n)]
    # Compute y^T*H*y
    yTHy = sum(y[i] * Hy[i] for i in range(n))

    H_new = [[0.0] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            H_new[i][j] = (
                H[i][j]
                - rho_val * (s[i] * yTH[j] + Hy[i] * s[j])
                + rho_val * (rho_val * yTHy + 1.0) * s[i] * s[j]
            )
    return H_new


def bfgs(
    f: Callable[[List[float]], float],
    x0: List[float],
    grad: Optional[Callable[[List[float]], List[float]]] = None,
    grad_tol: float = 1e-8,
    step_tol: float = 1e-8,
    func_tol: float = 1e-12,
    max_iterations: int = 1000,
    **kwargs,
) -> OptimizeResult:
    """Minimize using BFGS quasi-Newton method."""
    opts = default_options(grad_tol=grad_tol, step_tol=step_tol, func_tol=func_tol,
                           max_iterations=max_iterations)
    grad_fn = grad if grad is not None else (lambda x: forward_diff_gradient(f, x))

    n = len(x0)
    x = x0[:]
    fx = f(x)
    gx = grad_fn(x)
    function_calls = 1
    gradient_calls = 1

    # Check initial convergence
    grad_norm = norm_inf(gx)
    reason = check_convergence(grad_norm, float('inf'), float('inf'), 0, opts)
    if reason and is_converged(reason):
        return OptimizeResult(
            x=x[:], fun=fx, gradient=gx[:],
            iterations=0, function_calls=function_calls,
            gradient_calls=gradient_calls, converged=True,
            message=convergence_message(reason),
        )

    H = identity_matrix(n)

    for iteration in range(1, opts.max_iterations + 1):
        d = negate(mat_vec_mul(H, gx))

        ls = wolfe_line_search(f, grad_fn, x, d, fx, gx)
        function_calls += ls.function_calls
        gradient_calls += ls.gradient_calls

        if not ls.success:
            return OptimizeResult(
                x=x[:], fun=fx, gradient=gx[:],
                iterations=iteration, function_calls=function_calls,
                gradient_calls=gradient_calls, converged=False,
                message="Stopped: line search failed",
            )

        x_new = add_scaled(x, d, ls.alpha)
        f_new = ls.f_new
        g_new = ls.g_new if ls.g_new is not None else grad_fn(x_new)
        if ls.g_new is None:
            gradient_calls += 1

        sk = sub(x_new, x)
        yk = sub(g_new, gx)
        ys = dot(yk, sk)

        if ys > 1e-10:
            rho_val = 1.0 / ys
            H = bfgs_update(H, sk, yk, rho_val)

        step_norm = norm_inf(sk)
        func_change = abs(fx - f_new)
        grad_norm = norm_inf(g_new)

        x = x_new
        fx = f_new
        gx = g_new

        reason = check_convergence(grad_norm, step_norm, func_change, iteration, opts)
        if reason:
            return OptimizeResult(
                x=x[:], fun=fx, gradient=gx[:],
                iterations=iteration, function_calls=function_calls,
                gradient_calls=gradient_calls,
                converged=is_converged(reason),
                message=convergence_message(reason),
            )

    return OptimizeResult(
        x=x[:], fun=fx, gradient=gx[:],
        iterations=opts.max_iterations, function_calls=function_calls,
        gradient_calls=gradient_calls, converged=False,
        message=f"Stopped: reached maximum iterations ({opts.max_iterations})",
    )
